gtsfm/bundle/two_step_ba.py

```diff
@@ -0,0 +1,97 @@
+                                                         after the 1st round of BA. Defaults to
+                                                         DEFAULT_INTERMEDIATE_REPROJ_ERROR_THRESH.
+            output_reproj_error_thresh (optional): Reprojection error threshold for the final result. Defaults to None.
+            robust_measurement_noise (optional): Flag to enable use of robust noise model for measurement noise.
+                                                 Defaults to False.
+            shared_calib (optional): Flag to enable shared calibration across all cameras. Defaults to False.
+            max_iterations (optional): Max number of iterations when optimizing the factor graph. None means no cap.
+                                       Defaults to None.
+        """
+        self._ba1 = BundleAdjustmentOptimizer(
+            output_reproj_error_thresh=intermediate_reproj_error_thresh,
+            robust_measurement_noise=robust_measurement_noise,
+            shared_calib=shared_calib,
+            max_iterations=FIRST_BA_NUM_ITERATIONS,
+        )
+
+        super().__init__(
+            output_reproj_error_thresh=output_reproj_error_thresh,
+            robust_measurement_noise=robust_measurement_noise,
+            shared_calib=shared_calib,
+            max_iterations=None if max_iterations is None else max(1, max_iterations - FIRST_BA_NUM_ITERATIONS),
+        )
+
+    def run(
+        self,
+        initial_data: GtsfmData,
+        absolute_pose_priors: List[Optional[PosePrior]],
+        relative_pose_priors: Dict[Tuple[int, int], PosePrior],
+        verbose: bool = True,
+    ) -> Tuple[GtsfmData, GtsfmData, List[bool]]:
+        """Run the bundle adjustment by forming factor graph and optimizing using Levenberg–Marquardt optimization.
+
+        Args:
+            initial_data: initialized cameras, tracks w/ their 3d landmark from triangulation.
+            absolute_pose_priors: priors to be used on cameras.
+            relative_pose_priors: priors on the pose between two cameras.
+            verbose: Boolean flag to print out additional info for debugging.
+
+        Results:
+            Optimized camera poses, 3D point w/ tracks, and error metrics.
+            Optimized camera poses after filtering landmarks (and cameras with no remaining landmarks).
+            Valid mask as a list of booleans, indicating for each input track whether it was below the re-projection
+                threshold.
+        """
+        _, intermediate_filtered_result, mask = self._ba1.run(
+            initial_data=initial_data,
+            absolute_pose_priors=absolute_pose_priors,
+            relative_pose_priors=relative_pose_priors,
+            verbose=verbose,
+        )
+        valid_tracks_for_2nd_step = np.where(mask)[0]
+
+        final_result_unfiltered, final_result_filtered, mask = super().run(
```

Comment:
Looks good. Could you verify if >2 rounds helps even more?

Reply:
If so, we could generalize `TwoStepBA` to `MultiStepBA`.

Reply:
I think we decided to do this in a different PR, did not want to block this one for long.

Reply:
I had the same comment :)

Reply:
Yep I think we should do some experiments for this, but I didn't have time to go through with iterative BA iterations -> filtering